File: 0003-longest-substring-without-repeating-characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        map<char, int> mp;
        int maxLength = 0, start = 0;

        for (int i = 0; i < s.length(); i++) {
            char current = s[i];
            if (mp.find(current) != mp.end() && mp[current] >= start) {
                start = mp[current] + 1;
            }
            mp[current] = i;
            maxLength = max(maxLength, i - start + 1);
        }

        return maxLength;
    }
};

File: 0003-longest-substring-without-repeating-characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int max = 0;
        int left = 0;

        for (int right = 0; right < s.length(); right++) {
            char currentChar = s.charAt(right);
            if (map.containsKey(currentChar)) {
                left = Math.max(left, map.get(currentChar) + 1);
            }
            
            map.put(currentChar, right);
            max = Math.max(max, right - left + 1);
        }

        return max;
    }
}

